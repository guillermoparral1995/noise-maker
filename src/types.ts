import type { ActionTypes as EnvelopeActionTypes } from './components/Controls/EnvelopeControls/store/actions';
import type { ActionTypes as GeneralControlsActionTypes } from './components/Controls/GeneralControls/store/actions';
import type { ActionTypes as FilterActionTypes } from './components/Controls/FilterControls/store/actions';
import type { ActionTypes as LFOActionTypes } from './components/Controls/LFOControls/store/actions';
import { LFOTargetAction } from './providers/AudioContextProvider';

export type ActionTypes =
  | EnvelopeActionTypes
  | GeneralControlsActionTypes
  | FilterActionTypes
  | LFOTargetAction
  | LFOActionTypes;

export enum Waveform {
  SINE = 'sine',
  SQUARE = 'square',
  SAWTOOTH = 'sawtooth',
  TRIANGLE = 'triangle',
}

export enum FilterType {
  LOWPASS = 'lowpass',
  HIGHPASS = 'highpass',
}

export enum Selectors {
  WAVEFORM = 'waveform',
  FILTER = 'filter',
  LFO_TARGET = 'lfoTarget',
  LFO_WAVEFORM = 'lfoWaveform',
}

export enum Knobs {
  VOLUME = 'VOLUME',
  PAN = 'PAN',
  DETUNE = 'DETUNE',
  ATTACK = 'ATTACK',
  DECAY = 'DECAY',
  SUSTAIN = 'SUSTAIN',
  RELEASE = 'RELEASE',
  FILTER_CUTOFF = 'FILTER_CUTOFF',
  FILTER_RESONANCE = 'FILTER_RESONANCE',
  LFO_FREQUENCY = 'LFO_FREQUENCY',
  LFO_AMPLITUDE = 'LFO_AMPLITUDE',
}

export type LFOTarget =
  | Exclude<
      Knobs,
      | Knobs.LFO_FREQUENCY
      | Knobs.LFO_AMPLITUDE
      | Knobs.ATTACK
      | Knobs.DECAY
      | Knobs.SUSTAIN
      | Knobs.RELEASE
    >
  | 'off';

export interface State {
  volume: number;
  pan: number;
  waveform: Waveform;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  filter: {
    type: FilterType;
    cutoff: number;
    resonance: number;
  };
  lfo: {
    target: LFOTarget;
    waveform: Waveform;
    frequency: number;
    amplitude: number;
  };
}

export enum Actions {
  UPDATE_VOLUME,
  UPDATE_PAN,
  UPDATE_WAVEFORM,
  UPDATE_DETUNE,
  UPDATE_ATTACK,
  UPDATE_DECAY,
  UPDATE_SUSTAIN,
  UPDATE_RELEASE,
  UPDATE_FILTER_TYPE,
  UPDATE_FILTER_CUTOFF,
  UPDATE_FILTER_RESONANCE,
  UPDATE_LFO_TARGET,
  UPDATE_LFO_WAVEFORM,
  UPDATE_LFO_FREQUENCY,
  UPDATE_LFO_AMPLITUDE,
}
export type ActionBuilder<T> = (payload: T) => ActionTypes;
