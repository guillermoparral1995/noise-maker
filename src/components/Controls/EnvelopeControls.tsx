import React from 'react';
import Knob from '../shared/Knob';
import {
  updateAttack,
  updateDecay,
  updateRelease,
  updateSustain,
} from '../../store/actions';

const EnvelopeFilterControls = () => {
  return (
    <div>
      <Knob
        label="attack"
        defaultValue={0}
        action={updateAttack}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="decay"
        defaultValue={0}
        action={updateDecay}
        min={0}
        max={5}
      ></Knob>
      <Knob
        label="sustain"
        defaultValue={1}
        action={updateSustain}
        min={0}
        max={1}
      ></Knob>
      <Knob
        label="release"
        defaultValue={0}
        action={updateRelease}
        min={0}
        max={5}
      ></Knob>
    </div>
  );
};

export default EnvelopeFilterControls;
