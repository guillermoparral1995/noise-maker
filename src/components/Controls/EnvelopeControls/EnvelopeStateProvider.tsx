import React, { PropsWithChildren, useReducer } from 'react';
import initialState, { EnvelopeState } from './store/initialState';
import { ActionTypes } from './store/actions';
import reducer from './store/reducer';

interface EnvelopeStateProviderValue {
  state: EnvelopeState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const envelopeStateContext =
  React.createContext<EnvelopeStateProviderValue>(undefined);

export const EnvelopeStateProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <envelopeStateContext.Provider value={{ state, dispatch }}>
      {children}
    </envelopeStateContext.Provider>
  );
};
