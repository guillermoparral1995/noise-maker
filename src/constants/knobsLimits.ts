import { Knobs } from '../types';

interface KnobLimit {
  min: number;
  max: number;
  default: number;
}

export const knobsLimits: Record<Knobs, KnobLimit> = {
  [Knobs.VOLUME]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
  [Knobs.PAN]: {
    min: -1,
    max: 1,
    default: 0,
  },
  [Knobs.ATTACK]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.DECAY]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.SUSTAIN]: {
    min: 0,
    max: 1,
    default: 1,
  },
  [Knobs.RELEASE]: {
    min: 0,
    max: 5,
    default: 0,
  },
  [Knobs.FILTER_CUTOFF]: {
    min: 10,
    max: 5000,
    default: 350,
  },
  [Knobs.FILTER_RESONANCE]: {
    min: 0.0001,
    max: 50,
    default: 1,
  },
  [Knobs.LFO_FREQUENCY]: {
    min: 0.2,
    max: 10,
    default: 1,
  },
  [Knobs.LFO_AMPLITUDE]: {
    min: 0,
    max: 1,
    default: 0.5,
  },
};
