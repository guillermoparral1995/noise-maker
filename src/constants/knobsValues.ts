import {
  updateAttack,
  updateDecay,
  updateDetune,
  updatePitchbend,
  updateRelease,
  updateSustain,
} from '../components/Controls/EnvelopeControls/store/actions';
import {
  updateFilterCutoff,
  updateFilterResonance,
} from '../components/Controls/FilterControls/store/actions';
import {
  updatePan,
  updateVolume,
} from '../components/Controls/GeneralControls/store/actions';
import {
  updateLFO1Amplitude,
  updateLFO1Frequency,
  updateLFO2Amplitude,
  updateLFO2Frequency,
} from '../components/Controls/LFOControls/store/actions';
import { ActionBuilder, Knobs } from '../types';

interface KnobValue {
  label: string;
  min: number;
  max: number;
  default: number;
  midiControl?: number;
  action: ActionBuilder<number>;
  step?: number;
}

export const knobsValues: Record<Knobs, KnobValue> = {
  [Knobs.VOLUME]: {
    label: 'Volume',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 74,
    action: updateVolume,
  },
  [Knobs.PAN]: {
    label: 'Pan',
    min: -1,
    max: 1,
    default: 0,
    midiControl: 18,
    action: updatePan,
  },
  [Knobs.DETUNE]: {
    label: 'Detune',
    min: -1200,
    max: 1200,
    default: 0,
    midiControl: 75,
    action: updateDetune,
    step: 1,
  },
  [Knobs.PITCHBEND]: {
    label: 'Pitchbend',
    min: -300,
    max: 300,
    default: 0,
    action: updatePitchbend,
    step: 1,
  },
  [Knobs.ATTACK]: {
    label: 'Attack',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 93,
    action: updateAttack,
  },
  [Knobs.DECAY]: {
    label: 'Decay',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 91,
    action: updateDecay,
  },
  [Knobs.SUSTAIN]: {
    label: 'Sustain',
    min: 0,
    max: 1,
    default: 1,
    midiControl: 73,
    action: updateSustain,
  },
  [Knobs.RELEASE]: {
    label: 'Release',
    min: 0,
    max: 5,
    default: 0,
    midiControl: 79,
    action: updateRelease,
  },
  [Knobs.FILTER_CUTOFF]: {
    label: 'Cut-off',
    min: 10,
    max: 5000,
    default: 350,
    midiControl: 71,
    action: updateFilterCutoff,
    step: 1,
  },
  [Knobs.FILTER_RESONANCE]: {
    label: 'Resonance',
    min: 0.0001,
    max: 50,
    default: 1,
    midiControl: 19,
    action: updateFilterResonance,
    step: 0.1,
  },
  [Knobs.LFO_1_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 76,
    action: updateLFO1Frequency,
  },
  [Knobs.LFO_1_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 16,
    action: updateLFO1Amplitude,
  },
  [Knobs.LFO_2_FREQUENCY]: {
    label: 'Frequency',
    min: 0.2,
    max: 10,
    default: 1,
    midiControl: 77,
    action: updateLFO2Frequency,
  },
  [Knobs.LFO_2_AMPLITUDE]: {
    label: 'Amplitude',
    min: 0,
    max: 1,
    default: 0.5,
    midiControl: 17,
    action: updateLFO2Amplitude,
  },
};
