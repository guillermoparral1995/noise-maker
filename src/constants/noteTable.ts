export interface NoteValue {
  frequency: number;
  mapping: string;
}

type NoteTable = Record<Notes, NoteValue>;

export enum Notes {
  C3 = 'C3',
  CSharp3 = 'C#3',
  D3 = 'D3',
  DSharp3 = 'D#3',
  E3 = 'E3',
  F3 = 'F3',
  FSharp3 = 'F#3',
  G3 = 'G3',
  GSharp3 = 'G#3',
  A3 = 'A3',
  ASharp3 = 'A#3',
  B3 = 'B3',
  C4 = 'C4',
  CSharp4 = 'C#4',
  D4 = 'D4',
  DSharp4 = 'D#4',
  E4 = 'E4',
  F4 = 'F4',
  FSharp4 = 'F#4',
  G4 = 'G4',
  GSharp4 = 'G#4',
  A4 = 'A4',
  ASharp4 = 'A#4',
  B4 = 'B4',
  C5 = 'C5',
}

const noteTable: NoteTable = {
  [Notes.C3]: {
    frequency: 130.81,
    mapping: 'q',
  },
  [Notes.CSharp3]: {
    frequency: 138.59,
    mapping: '2',
  },
  [Notes.D3]: {
    frequency: 146.83,
    mapping: 'w',
  },
  [Notes.DSharp3]: {
    frequency: 155.56,
    mapping: '3',
  },
  [Notes.E3]: {
    frequency: 164.81,
    mapping: 'e',
  },
  [Notes.F3]: {
    frequency: 174.61,
    mapping: 'r',
  },
  [Notes.FSharp3]: {
    frequency: 185,
    mapping: '5',
  },
  [Notes.G3]: {
    frequency: 196,
    mapping: 't',
  },
  [Notes.GSharp3]: {
    frequency: 207.65,
    mapping: '6',
  },
  [Notes.A3]: {
    frequency: 220,
    mapping: 'y',
  },
  [Notes.ASharp3]: {
    frequency: 233.08,
    mapping: '7',
  },
  [Notes.B3]: {
    frequency: 246.94,
    mapping: 'u',
  },
  [Notes.C4]: {
    frequency: 261.63,
    mapping: 'z',
  },
  [Notes.CSharp4]: {
    frequency: 277.18,
    mapping: 's',
  },
  [Notes.D4]: {
    frequency: 293.66,
    mapping: 'x',
  },
  [Notes.DSharp4]: {
    frequency: 311.13,
    mapping: 'd',
  },
  [Notes.E4]: {
    frequency: 329.63,
    mapping: 'c',
  },
  [Notes.F4]: {
    frequency: 349.23,
    mapping: 'v',
  },
  [Notes.FSharp4]: {
    frequency: 369.99,
    mapping: 'g',
  },
  [Notes.G4]: {
    frequency: 392,
    mapping: 'b',
  },
  [Notes.GSharp4]: {
    frequency: 415.3,
    mapping: 'h',
  },
  [Notes.A4]: {
    frequency: 440,
    mapping: 'n',
  },
  [Notes.ASharp4]: {
    frequency: 466.16,
    mapping: 'j',
  },
  [Notes.B4]: {
    frequency: 493.88,
    mapping: 'm',
  },
  [Notes.C5]: {
    frequency: 523.25,
    mapping: ',',
  },
};

export default noteTable;
