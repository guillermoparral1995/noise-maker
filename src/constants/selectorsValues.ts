import {
  updateLFO1Target,
  updateLFO2Target,
} from '@providers/AudioContextProvider/store/actions';
import { updateMIDIInput } from '@providers/MIDIProvider/store/actions';
import {
  ActionBuilder,
  FilterType,
  Knobs,
  LFO1Target,
  LFO2Target,
  Selectors,
  Waveform,
} from '@types';
import { updateWaveform } from '../components/Controls/EnvelopeControls/store/actions';
import { updateFilterType } from '../components/Controls/FilterControls/store/actions';
import {
  updateLFO1Waveform,
  updateLFO2Waveform,
} from '../components/Controls/LFOControls/store/actions';

type DropdownValue = Waveform | FilterType | LFO1Target | LFO2Target | string;

export interface DropdownOption {
  label: string;
  value: DropdownValue;
}

interface SelectorValues {
  label: string;
  action: ActionBuilder<DropdownValue>;
  options: DropdownOption[];
}

const waveforms: DropdownOption[] = [
  {
    label: 'Sine',
    value: Waveform.SINE,
  },
  {
    label: 'Square',
    value: Waveform.SQUARE,
  },
  {
    label: 'Sawtooth',
    value: Waveform.SAWTOOTH,
  },
  {
    label: 'Triangle',
    value: Waveform.TRIANGLE,
  },
];

const filterTypes: DropdownOption[] = [
  {
    label: 'Low-pass',
    value: FilterType.LOWPASS,
  },
  {
    label: 'High-pass',
    value: FilterType.HIGHPASS,
  },
];

const lfo1Targets: DropdownOption[] = [
  {
    label: 'Off',
    value: 'off',
  },
  {
    label: 'Volume',
    value: Knobs.VOLUME,
  },
  {
    label: 'Pan',
    value: Knobs.PAN,
  },
  {
    label: 'Cut-off',
    value: Knobs.FILTER_CUTOFF,
  },
  {
    label: 'Resonance',
    value: Knobs.FILTER_RESONANCE,
  },
  {
    label: 'Detune',
    value: Knobs.DETUNE,
  },
];

export const selectorValues: Record<Selectors, SelectorValues> = {
  [Selectors.WAVEFORM]: {
    label: 'Waveform',
    action: updateWaveform,
    options: waveforms,
  },
  [Selectors.LFO_1_WAVEFORM]: {
    label: 'Waveform',
    action: updateLFO1Waveform,
    options: waveforms,
  },
  [Selectors.LFO_2_WAVEFORM]: {
    label: 'Waveform',
    action: updateLFO2Waveform,
    options: waveforms,
  },
  [Selectors.FILTER]: {
    label: 'Type',
    action: updateFilterType,
    options: filterTypes,
  },
  [Selectors.LFO_1_TARGET]: {
    label: 'Target',
    action: updateLFO1Target,
    options: lfo1Targets,
  },
  [Selectors.LFO_2_TARGET]: {
    label: 'Target',
    action: updateLFO2Target,
    options: [
      ...lfo1Targets,
      {
        label: 'LFO 1 Frequency',
        value: Knobs.LFO_1_FREQUENCY,
      },
      {
        label: 'LFO 1 Amplitude',
        value: Knobs.LFO_1_AMPLITUDE,
      },
    ],
  },
  [Selectors.MIDI_INPUT]: {
    label: 'MIDI Input',
    action: updateMIDIInput,
    options: [],
  },
};
