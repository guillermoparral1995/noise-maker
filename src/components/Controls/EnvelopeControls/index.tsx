import useAddMidiListeners from '@hooks/useAddMidiListeners';
import { Knobs } from '@types';
import React, { useContext } from 'react';
import ControlsRow from '../../shared/ControlsRow';
import Knob from '../../shared/Knob';
import { envelopeStateContext } from './EnvelopeStateProvider';

const EnvelopeControls = () => {
  const { state, dispatch } = useContext(envelopeStateContext);
  useAddMidiListeners(
    [Knobs.ATTACK, Knobs.DECAY, Knobs.SUSTAIN, Knobs.RELEASE],
    dispatch,
  );

  return (
    <>
      <h3>ADSR Envelope</h3>
      <ControlsRow>
        <Knob id={Knobs.ATTACK} value={state.attack} dispatch={dispatch}></Knob>
        <Knob id={Knobs.DECAY} value={state.decay} dispatch={dispatch}></Knob>
        <Knob
          id={Knobs.SUSTAIN}
          value={state.sustain}
          dispatch={dispatch}
        ></Knob>
        <Knob
          id={Knobs.RELEASE}
          value={state.release}
          dispatch={dispatch}
        ></Knob>
      </ControlsRow>
    </>
  );
};

export default EnvelopeControls;
