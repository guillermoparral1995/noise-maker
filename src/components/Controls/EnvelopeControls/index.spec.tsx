import { cleanup, render } from '@testing-library/react';
import React from 'react';
import EnvelopeControls from '.';
import { knobsValues } from '../../../constants/knobsValues';
import {
  withMockedMIDIInput,
  withMockedMIDINoInput,
} from '../../../providers/MIDIProvider';
import { Actions, Knobs } from '../../../types';
import { EnvelopeStateProvider } from './EnvelopeStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

describe('Envelope controls', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render correctly', async () => {
    const { findByTestId } = render(
      withMockedMIDINoInput(
        <EnvelopeStateProvider>
          <div data-testid="envelope-controls">
            <EnvelopeControls></EnvelopeControls>
          </div>
        </EnvelopeStateProvider>,
      ),
    );

    const controls = await findByTestId('envelope-controls');
    expect(controls).toMatchSnapshot();
  });

  it('should trigger dispatch on MIDI knob change', async () => {
    const mockDispatch = jest.fn();
    let mockEventName: string;
    const mockAddListener = jest.fn(
      (eventName: string, listener: (e: object) => void) => {
        mockEventName = eventName;
        listener({
          value: 0.5,
          controller: { number: knobsValues[Knobs.ATTACK].midiControl },
        });
      },
    );

    __mockGetInputByName.mockReturnValue({
      addListener: mockAddListener,
    });

    const { findByTestId } = render(
      withMockedMIDIInput(
        <EnvelopeStateProvider __mockDispatch={mockDispatch}>
          <EnvelopeControls></EnvelopeControls>
        </EnvelopeStateProvider>,
      ),
    );

    await findByTestId(Knobs.ATTACK);
    expect(mockEventName).toEqual('controlchange');
    expect(mockDispatch).toHaveBeenCalledWith({
      type: Actions.UPDATE_ATTACK,
      payload: 2.5,
    });
  });
});
