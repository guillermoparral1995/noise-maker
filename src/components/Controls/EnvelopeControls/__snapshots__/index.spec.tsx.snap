// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Envelope controls should render correctly 1`] = `
<div
  data-testid="envelope-controls"
>
  <h3>
    ADSR Envelope
  </h3>
  <div
    class="controls_row"
  >
    <div
      data-testid="ATTACK"
      tabindex="0"
    >
      <label
        class="label"
        id="knob_label_ATTACK"
      >
        Attack
      </label>
      <div
        aria-labelledby="knob_label_ATTACK"
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="ATTACK"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 29.999999999999982 84.64101615137754"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="ATTACK"
            text-anchor="middle"
            x="50"
            y="57"
          >
            0
          </text>
        </svg>
      </div>
    </div>
    <div
      data-testid="DECAY"
      tabindex="0"
    >
      <label
        class="label"
        id="knob_label_DECAY"
      >
        Decay
      </label>
      <div
        aria-labelledby="knob_label_DECAY"
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="DECAY"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 29.999999999999982 84.64101615137754"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="DECAY"
            text-anchor="middle"
            x="50"
            y="57"
          >
            0
          </text>
        </svg>
      </div>
    </div>
    <div
      data-testid="SUSTAIN"
      tabindex="0"
    >
      <label
        class="label"
        id="knob_label_SUSTAIN"
      >
        Sustain
      </label>
      <div
        aria-labelledby="knob_label_SUSTAIN"
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="SUSTAIN"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70.00000000000001 84.64101615137754"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="SUSTAIN"
            text-anchor="middle"
            x="50"
            y="57"
          >
            1
          </text>
        </svg>
      </div>
    </div>
    <div
      data-testid="RELEASE"
      tabindex="0"
    >
      <label
        class="label"
        id="knob_label_RELEASE"
      >
        Release
      </label>
      <div
        aria-labelledby="knob_label_RELEASE"
        class="p-knob p-component"
        data-pc-name="knob"
        data-pc-section="root"
        id="RELEASE"
      >
        <svg
          data-pc-section="svg"
          height="50"
          viewBox="0 0 100 100"
          width="50"
        >
          <path
            class="p-knob-range"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 1 1 70 84.64101615137754"
            data-pc-section="range"
            stroke="var(--surface-border, LightGray)"
            stroke-width="5"
          />
          <path
            class="p-knob-value"
            d="M 29.999999999999982 84.64101615137754 A 40 40 0 0 1 29.999999999999982 84.64101615137754"
            data-pc-section="value"
            stroke="var(--primary-color, Black)"
            stroke-width="5"
          />
          <text
            class="p-knob-text"
            data-pc-section="label"
            fill="var(--text-color-secondary, Black)"
            name="RELEASE"
            text-anchor="middle"
            x="50"
            y="57"
          >
            0
          </text>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
